blink/procfs.c

```diff
@@ -16,94 +16,1463 @@
+                           .readlink = ProcfsToSelfReadlink},
+    [PROCFS_SELF_INO] = {PROCFS_SELF_INO, S_IFLNK | 0777, 0, 0,
+                         PROCFS_SELF_TYPE, "self",
+                         .readlink = ProcfsSelfReadlink},
+    [PROCFS_SYS_INO] = {PROCFS_SYS_INO, S_IFDIR | 0555, 0, 0, PROCFS_SYS_TYPE,
+                        "sys"},
+    [PROCFS_UPTIME_INO] = {PROCFS_UPTIME_INO, S_IFREG | 0444, 0, 0,
+                           PROCFS_UPTIME_TYPE, "uptime",
+                           .read = ProcfsUptimeRead},
+};
+
+static struct ProcfsInfo g_piddirinfos[] = {
+    [PROCFS_PIDDIR_TYPE -
+     PROCFS_PIDDIR_TYPE] = {0, S_IFDIR | 0555, 0, 0, PROCFS_PIDDIR_TYPE, "",
+                            .readdir = ProcfsPiddirReaddir},
+    [PROCFS_PIDDIR_EXE_TYPE -
+        PROCFS_PIDDIR_TYPE] = {0, S_IFLNK | 0777, 0, 0, PROCFS_PIDDIR_EXE_TYPE,
+                               "exe", .readlink = ProcfsPiddirExeReadlink},
+    [PROCFS_PIDDIR_CWD_TYPE -
+        PROCFS_PIDDIR_TYPE] = {0, S_IFLNK | 0777, 0, 0, PROCFS_PIDDIR_CWD_TYPE,
+                               "cwd", .readlink = ProcfsPiddirCwdReadlink},
+    [PROCFS_PIDDIR_ROOT_TYPE -
+        PROCFS_PIDDIR_TYPE] = {0, S_IFLNK | 0777, 0, 0, PROCFS_PIDDIR_ROOT_TYPE,
+                               "root", .readlink = ProcfsPiddirRootReadlink},
+    [PROCFS_PIDDIR_MOUNTS_TYPE -
+        PROCFS_PIDDIR_TYPE] = {0, S_IFREG | 0444, 0, 0,
+                               PROCFS_PIDDIR_MOUNTS_TYPE, "mounts",
+                               .read = ProcfsPiddirMountsRead},
+    [PROCFS_PIDDIR_FDDIR_TYPE - PROCFS_PIDDIR_TYPE] = {0, S_IFDIR | 0555, 0, 0,
+                                                       PROCFS_PIDDIR_FDDIR_TYPE,
+                                                       "fd"},
+};
+
+////////////////////////////////////////////////////////////////////////////////
+
+static int ProcfsCreateInfo(struct ProcfsInfo **info) {
+  *info = malloc(sizeof(struct ProcfsInfo));
+  if (*info == NULL) {
+    return enomem();
+  }
+  (*info)->ino = 0;
+  (*info)->mode = 0;
+  (*info)->type = 0;
+  (*info)->name[0] = '\0';
+  (*info)->openfile = NULL;
+  return 0;
+}
+
+static int ProcfsCreateDefaultInfo(struct ProcfsInfo **info,
+                                   struct ProcfsDevice *device, u64 ino) {
+  *info = malloc(sizeof(struct ProcfsInfo));
+  if (*info == NULL) {
+    return enomem();
+  }
+  **info = g_defaultinfos[ino];
+  (*info)->time = device->mounttime;
+  return 0;
+}
+
+static int ProcfsCreatePiddirInfo(struct ProcfsInfo **info,
+                                  struct ProcfsInfo *parent, i32 pid,
+                                  u32 type) {
+  *info = malloc(sizeof(struct ProcfsInfo));
+  if (*info == NULL) {
+    return enomem();
+  }
+  if (parent == NULL) {
+    **info = g_piddirinfos[0];
+    // TODO(trungnt): This should be dynamically allocated when we support
+    // more processes.
+    (*info)->ino = PROCFS_FIRST_PID_INO;
+    sprintf((*info)->name, "%d", pid);
```

Comment:
I recommend against using sprintf() because it's one of those functions that triggers security people and attracts attention from scanners.

Reply:
I used sprintf at that once instance as I wonder how a 32-bit int can take more than   16 chars to represent.